pkg/apis/core/validation/shoot.go

```diff
@@ -1699,27 +1699,31 @@ func validateShootOperation(operation string, shoot *core.Shoot, fldPath *field.
 
 	switch operation {
 	case v1beta1constants.ShootOperationRotateCAStart:
-		if shoot.Status.LastOperation == nil ||
-			(shoot.Status.LastOperation.Type != core.LastOperationTypeReconcile &&
-				(shoot.Status.LastOperation.Type != core.LastOperationTypeCreate || shoot.Status.LastOperation.State != core.LastOperationStateSucceeded)) {
-			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot start CA rotation if shoot was created successfully or is ready for reconciliation"))
+		if !isShootReadyForCARotationStart(shoot.Status.LastOperation) {
+			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot start CA rotation if shoot was not yet created successfully or is not ready for reconciliation"))
 		}
-
-		if shoot.Status.Credentials != nil &&
-			shoot.Status.Credentials.Rotation != nil &&
-			shoot.Status.Credentials.Rotation.CertificateAuthorities != nil &&
-			shoot.Status.Credentials.Rotation.CertificateAuthorities.Phase != core.RotationCompleted {
+		if phase := helper.GetShootCARotationPhase(shoot.Status.Credentials); len(phase) > 0 && phase != core.RotationCompleted {
 			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot start CA rotation if .status.credentials.rotation.certificateAuthorities.phase is not 'Completed'"))
 		}
 
 	case v1beta1constants.ShootOperationRotateCAComplete:
-		if shoot.Status.Credentials == nil ||
-			shoot.Status.Credentials.Rotation == nil ||
-			shoot.Status.Credentials.Rotation.CertificateAuthorities == nil ||
-			shoot.Status.Credentials.Rotation.CertificateAuthorities.Phase != core.RotationPrepared {
+		if helper.GetShootCARotationPhase(shoot.Status.Credentials) != core.RotationPrepared {
 			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot complete CA rotation if .status.credentials.rotation.certificateAuthorities.phase is not 'Prepared'"))
 		}
 	}
 
 	return allErrs
 }
+
+func isShootReadyForCARotationStart(lastOperation *core.LastOperation) bool {
+	if lastOperation == nil {
+		return false
+	}
+	if lastOperation.Type == core.LastOperationTypeCreate && lastOperation.State == core.LastOperationStateSucceeded {
+		return true
+	}
+	if lastOperation.Type == core.LastOperationTypeRestore && lastOperation.State == core.LastOperationStateSucceeded {
```

Comment:
Sorry for my question but I couldn't quickly derive that from the code. Is it ok to rotate after a successful migrate operation?
`lastOperation.Type == core.LastOperationTypeMigrate && lastOperation.State == core.LastOperationStateSucceeded`

Reply:
Probably not, `Migrate Succeeded` means that the old control plane was just terminated and is about to be restored.

Reply:
That's what I thought as well but wanted to double check